### DISCUSSION TITLE
[PPR] TypeError: Response.clone: Body has already been consumed.

### DISCUSSION THREAD

**Discussion by steve-marmalade** [vercel/next.js#69635] [App Router] [answered]
@wyattjoh , @ztanner 

Testing out PPR and PFPR and it seems to be working well for the most part (nice work). However, in one of my routes I am seeing a 500 for the prefetch RSC GET request. Clicking a Link to those pages causes a full page load, but it does ultimately load the content.

```
⨯ TypeError: Response.clone: Body has already been consumed.
    at webidl.errors.exception (node:internal/deps/undici/undici:3384:14)
    at _Response.clone (node:internal/deps/undici/undici:8876:31)
    at /var/task/.next/server/chunks/8875.js:2:11741
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async /var/task/.next/server/chunks/8047.js:1:182603
    at async K (/var/task/.next/server/chunks/8047.js:1:184526)
    at async /var/task/.next/server/chunks/4498.js:166:33
    at async v (/var/task/.next/server/app/path/to/my/profile/page.js:1:4306) {
  digest: '1674423162',
  page: '/path/to/my/profile'
}
```

I'm not opening an Issue yet as I don't have a clear way to reproduce (I don't know what's different about this page from others that are working). Happy to help investigate this if you can give me some pointers on what to try / where to look. Also happy to share a Vercel preview build URL / create a Replay, if that's helpful.

**Comment by steve-marmalade** [comment]
Fixed in https://github.com/vercel/next.js/pull/70082 :tada:

**Reply by justin-mclaren** [reply]
We're dealing with this error in our production build on `next@14.2.9`. Is the only way to resolve it by upgrading to the canary release for v15?

**Reply by lapa182** [reply]
I saw this was backported to 14.2.14, but I can still see the same issue.

**Reply by justin-mclaren** [reply]
Ya I just tested out 14.2.14 but we're still seeing it as well.

**Reply by lapa182** [reply]
@steve-marmalade could you confirm to us?

**Reply by alexander-tsaplev-cko** [reply]
Same — I experience this error on 14.2.10

**Reply by rakhi2104** [reply]
we are facing same issue in 14.2.5 as well. 
@steve-marmalade what is a safe version to downgrade?

**Reply by steve-marmalade** [reply]
Hi folks, I'm afraid I don't have any additional context to provide (I'm just a user). I'm on the latest canaries and this issue has been resolved for me there.

**Reply by alexander-tsaplev-cko** [reply]
Looks like pinning the Docker version helped us solve the issue... It might be worth checking your environment

**Reply by justin-mclaren** [reply]
@alexander-tsaplev-cko what Docker version did you pin to?

**Reply by mogzol** [reply]
I'm on 14.2.15 and still seeing this. According to the error, it is occurring at `app-page.runtime.prod.js:12:181023`, which corresponds to this line: https://github.com/vercel/next.js/blob/v14.2.15/packages/next/src/server/lib/patch-fetch.ts#L738

Looking at the backport PR, that part wasn't modified at all, the changes are just after that: https://github.com/vercel/next.js/pull/70649

So I think the backport did not properly fix this issue unfortunately

**Reply by OZZlE** [reply]
@steve-marmalade can we expect the fix to be made on v14 stable version any time soon? Our entire website is currently broken, we have lots of logging in the app but this is the only reported error I can find

**Reply by mogzol** [reply]
@OZZlE should be fixed properly with [this PR](https://github.com/vercel/next.js/pull/71115) which is merged, so I would expect the next v14 release to include it

**Reply by rakhi2104** [reply]
hey @steve-marmalade,

We did get this working when we downgraded our NodeJS Docker image to 18-alpine and sticking to Next @ 14.2.15

**Reply by beckerei** [reply]
I could not observer the behaviour when downgrading to node `20.16` which does not throw the error together with next `14.2.14`. 
Node 20.17 contains a memory leak in fetch, which is fixed in 20.18. But 20.18 and also 22.9 will are effected by the problem. (we run node alpine as well)
I'm unsure how node version and the problem in next are connected. We are not using PPR at all.

**Reply by OZZlE** [reply]
why is this marked as fixed when the issue obviously still exists in latest stable version e.g. Next 14, sure it works in node.js 18 but then that should be required or patched in next

**Reply by paullewisn** [reply]
I am getting this error. Next v14.2.6 and Node v22.11.0.

**Reply by yerfinojul** [reply]
Same error on NextJS v14.2.15 and Node v22.11.0

**Comment by sannajammeh** [comment]
Appears this error is now in all our Vercel deployments. Happened overnight with no changes to git.

**Reply by Eliotboutet** [reply]
Same for us here. Do we have any visibility on the issue resolution ?

**Reply by sannajammeh** [reply]
Can confirm this issue is due to NodeJS 20.18 which was gradually rolled out as one by one our deployments stopped building. More specifically it's a bug inside Undici 6.19. 6.20 is fixes the issue but not included in Node 20.18. We were able to fix build behaviour by polyfilling fetch to Undici 6.20. 

Short term fix is to Downgrade affected projects to Node JS 18.x by specifying this in your engines field in package.json. This fixes the issue. Then we wait for a new NodeJS patch and Vercel update

**Reply by stormsson** [reply]
@sannajammeh how did you actually polyfill fetch to that version of Undici ?

**Reply by sannajammeh** [reply]
> @sannajammeh how did you actually polyfill fetch to that version of Undici ?

Install undici, then inside root layer on server

```tsx
import { fetch } from "undici";

globalThis.fetch = fetch
```

Do not run this code in production. This is only to confirm the problem is undici in node 20.17 and up

**Reply by wutupake** [reply]
I solved with a `npm i undici --save-dev` (it install the 6.20.1 version) on my project (**on local only**). 
When Vercel uses a version of node greater than 20.18.0 I will remove the reference to undici from my package.json.
Let me know if this sounds like a good solution for you as well, thanks

**UPDATE**: add undici as a devDependencies solve the problem only on local environment. To solve on production you have to use the 18.x version of node on Vercel (until the update)

**Comment by dejoma** [comment]
I am using Vercel but I would love to stick to node v20 somehow; can I pin the version on v20.16 somehow?
```json
{
  "engines": {
    "node": "20.16"
  }
}
```

Would that work in `package.json` ?

**Reply by sannajammeh** [reply]
Impossible I'm afraid

**Comment by stormsson** [comment]
As a workaround in our situation we just added updated undici to package.json devDependencies and this solved the issue for the moment

**Reply by wutupake** [reply]
Yes, sorry @stormsson, this solution works only on local environment. The real workaround is still use the 18.x version of Node in Vercel. 
(I'll update my previous comment.)

**Reply by polnetwork** [reply]
Moving to Node18 did not worked for me (selfhosted using docker)

**Reply by OZZlE** [reply]
> Moving to Node18 did not worked for me (selfhosted using docker)

You probably didn't change node version correctly then, you need to change it inside Dockerfile

**Reply by polnetwork** [reply]
> You probably didn't change node version correctly then, you need to change it inside Dockerfile

Thanks for the response, moved to node 18 using `FROM node:18-alpine AS base` in Dockerfile.
Also checked inside the POD that it's using 18 
`/app $ node -v
v18.20.4`

**Reply by OZZlE** [reply]
strange, that solved our issue directly, unforturtuneatly this error wasn't the root cause of our issue rather than an external api not responding properly, at least it removed confusing logs though :) no idea why you still see it then.. have you manually installed unduci to a newer version perhaps? then try to uninstall it so it uses the default perhaps

**Comment by RyanWarner** [comment]
This is still happening to me even after upgrading to NextJS v15.02 and adding undici v6.20.1 to devDependencies.

**Comment by royporter7** [comment]
This is happening to me on my production application. Next 14.2.17 deployed via SST v3. Unfortunately SST v3 seems to only allow configuring Nextjs sites to Node v20, but manually going to the AWS console and downgrading the site default function to Node 18 seems to clear the problem up, however subsequent deployments appear to override the manual change.
For SST v3 users users, adding the below to your NextjsSite construct will fix this.
```
transform: {
  server: {
    runtime: "nodejs18.x"
  }
}
```
Preferably Next 14 is updated to not have this error occur.

**Comment by polnetwork** [comment]
Happening to me (self hosted with docker) with NextJS 14.2.17 and node:20-alpine, also tried node:18-alpine and same happens.

**Comment by albertocubeddu** [comment]
This is happening to me on my production application. Next 14.2.17 deployed via Vercel.

Localhost works fine...

**Comment by albertocubeddu** [comment]
It's impossible for me to reproduce the issue on local, i'm using the same nodejs version that is used on Vercel; And the configuration for the middleware is literally copy/pasted from supabase (most of the time i get the clone body error, as supabase is not able to read the information). 

(Running on Node.js version: v20.18.0) -> Works perfect on localhost, but not in production.

Strangely enough, the error is easy to reproduce in production, when i deploy a new version (just changing the readme, it will trigger it) after a while the error rate drastically reduce; However you can imagine how happy are my customers seeing random errors pages.


Some of the errors so that can google can index it and brings more people to this page: 
`TypeError: Response.clone: Body has already been consumed.
    at webidl.errors.exception (node:internal/deps/undici/undici:3384:14)
    at _Response.clone (node:internal/deps/undici/undici:8883:31)
    at /var/task/node_modules/.pnpm/next@14.2.17_@babel+core@7.24.5_@opentelemetry+api@1.9.0_@playwright+test@1.48.2_babel-plugin_nqwl6ilrdpb4nd765bcr723oqy/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:181023
    at runNextTicks (node:internal/process/task_queues:60:5)
    at listOnTimeout (node:internal/timers:545:9)
    at process.processTimers (node:internal/timers:519:7)`

`e0 [AuthRetryableFetchError]: Response.clone: Body has already been consumed.
    at aa (/var/task/.next/server/chunks/3239.js:6:26826)
    at runNextTicks (node:internal/process/task_queues:60:5)
    at listOnTimeout (node:internal/timers:545:9)
    at process.processTimers (node:internal/timers:519:7)
    at async ae (/var/task/.next/server/chunks/3239.js:6:26507)
    at async /var/task/.next/server/chunks/3239.js:6:49220
    at async ay._useSession (/var/task/.next/server/chunks/3239.js:6:47353)
    at async ay._getUser (/var/task/.next/server/chunks/3239.js:6:49044)
    at async /var/task/.next/server/chunks/3239.js:6:48895
    at async /var/task/.next/server/chunks/3239.js:6:46633 {
  __isAuthError: true,
  status: 0,
  code: undefined,
  digest: '319384388'
}`

#Update 9 Nov - 2:36AM

- The environmental variable VERCEL_UNDICI didn't make any difference for the problem stated above;
- Rolling back to node18 seems to solve the problem. That's not a solution; However it seems to pin down the problem.

**Comment by alexshyba** [comment]
👋 hello,

I am a bit confused as to what the resolution was. I have a customer running on app router `14.2.16` and not using PPR and the workaround for setting node.js to 18 resolved the issue.

Has this been resolved in 15, can somebody please help clarify? Thank you so much!

**Comment by donhessman** [comment]
@steve-marmalade Yeah I see the fix was merged to canary. We are not quite ready to update 15 yet though, much less canary.

Is there any chance we can get this fix back-ported to a nextjs 14+ security patch like they did for those issues fixed in 14.2.17?

**Reply by donhessman** [reply]
Or is 14.2.17 meant to also contain this same fix?

**Comment by envious** [comment]
this needs to be reopened, it still happens

**Reply by jaeilnet** [reply]
me too..

**Reply by jsonfm** [reply]
same with Nextjs 15

**Reply by jaeilnet** [reply]
page router Would it be different if I used a page router?

**Reply by albertocubeddu** [reply]
In my case is not a next issue at all; So using the page route wouldn't help. 

Have you tried downgrading nodejs to 18.x? That's the fix i'm currently using and from 3 days ago i didn't see a single error.

**Reply by jaeilnet** [reply]
I downgraded Undici to v6.19 and tested it and it seems to be working well at the moment

**Comment by tomi-mercado** [comment]
I am still having the same issue, and I can't downgrade to Node 18 because I am using the `File` class who doesn't exists there. Pin a previous version of Node 20 it's also not possible in Vercel settings.

**Comment by silviusmaximus** [comment]
Yeah we're running into this as well a bunch. I think its an underlying call to supabase which is failing, which is resulting in middleware saying there's no user and redirecting them to login page. Really annoying.

**Reply by vbrantner** [reply]
have the exact same problem with supabase

**Reply by tomi-mercado** [reply]
Same problem but with `next-auth` and Salesforce. When the error happens and I read the `session`, this one doesn't exists and causes a redirect to the login page.

**Comment by MattJustMatt** [comment]
Happening for me on NextJS 15 without (AFAIK) failed calls. Revert to Node 18 resolved.

**Comment by stefan-girlich** [comment]
Issue happens for me on Vercel only, using @supabase/ssr 0.5.1, @supabase/supabase-js 2.46.1, next 14.2.18.
In my case, the issue occurs only when trying to render pages, but not in the middleware. I created a discussion in the Supabase organization for visibility: https://github.com/orgs/supabase/discussions/30522

@steve-marmalade There are reports in the comments that this still occurs in Next.js 15, and even if there is a fix for 15.x, not everyone is able to migrate right away. Could you re-open this this issue or provider further guidance?

**Comment by MuhammadBilal1234** [comment]
How do we handle this error for now? Just create an error page and refresh the page when this error occurs?

**Comment by wesleyjanse** [comment]
This should really be re-openend since the solution that is marked is not the general solution for everyone on next 14. 
This issue is not fixed if we are forced to revert to node 18. How can this still not be solved, a lot of live production apps are running on these versions and end-users are impacted. They should really stop with marking all "rc" versions "production-ready"

**Comment by grzegorzpokorski** [comment]
Also encounter the following error randomly: `TypeError: Response.clone: Body has already been consumed.` This happens especially when I visit a dynamic page (SSR), such as `/blog/[slug]`. I am using Next.js v15.0.3, and the page is built and deployed on Vercel (Node.js 20.x). This error appears only after build an app not on dev mode.

**Reply by grzegorzpokorski** [reply]
In console I receive error like this:
```
 ⨯ TypeError: Response.clone: Body has already been consumed.
    at <unknown> (/home/name/project/.next/server/chunks/1241.js:2:6811) {
  digest: '217859647'
}
```
I spend time to find what code causes that error in my case and i found that `/home/name/project/.next/server/chunks/1241.js:2:6811` points to that code: https://github.com/vercel/next.js/blob/697aff9a4a156b332bef3da453273f3f355b9b79/packages/next/src/server/lib/dedupe-fetch.ts#L92

https://github.com/vercel/next.js/issues/73035

**Comment by yuki-and** [comment]
I am still having the same issue. I'm using node 18 and nextjs v14.2.11 and I'd like to solve this issue without downgrading to node 18.

**Comment by yp717** [comment]
This is still happening on v14.2.18 - is there any update on this?

Still getting: `Response.clone: Body has already been consumed.`

Fully blocking on next 14 / adopting app directory or deploying anything to production...

**Comment by maxpagels** [comment]
Same issue on 14.2.8, downgrading to node 18 works.

**Comment by paperdiamond** [comment]
Agreed that this should still be open - was experiencing this in next15 and dropped back to next14. Still experiencing it in both. 

Switching the node version to node 18 in vercel fixed it for me.

**Comment by ApacheEx** [comment]
I was using node20-alpine and nextjs v14.2.15, This error occurs on production already 55 times for last 7 days. It's insane, can you guys have a look on it with priority? thanks in advance

<img width="696" alt="image" src="https://github.com/user-attachments/assets/d2ff9986-0e8a-47ce-9f4c-24ddca190440">

**Comment by sannajammeh** [comment]
@ztanner 

Hoping we can get some attention directed to this. This is not a resolved conversation.
This is not a PPR issue, its a direct Next.js 14 issue

**Reply by grzegorzpokorski** [reply]
In my project I also get this error. I use next v15.0.3. In my case error in console points to the following line in code: https://github.com/vercel/next.js/blob/697aff9a4a156b332bef3da453273f3f355b9b79/packages/next/src/server/lib/dedupe-fetch.ts#L92

So I think that this issue is related to deduplication of fetch requests.

**Comment by wesleyjanse** [comment]
We have a Vercel partnership at work, and I had a chat with support. They suggest to set the following environment variable on Vercel. 

`VERCEL_UNDICI=1`

It seems to resolve the issue for us using `Node 20.x` and `Next 14.2.18`

Using that environment variable will use a different Undici version then Node uses. But I'm afraid this is only really a solution if you are hosting on Vercel.

**Comment by davyike** [comment]
We're not hosting on Vercel. We only saw this issue appear after we upgraded from Node 20.16 to 20.18.1. The timing of the initial comment in this thread coincides with (or slightly after) the release of Node 20.17. Is it possible a regression was introduced there, rather than in Next? Or an interaction between the two? I'd be curious to see someone else downgrade Node to 20.16 and see if the issue persists.

**Reply by sannajammeh** [reply]
It's a node 20.17 issue. Undici dependency introduced a breaking change, they have since reverted it but Vercel has not upgraded their build containers to latest node.

**Reply by yerfinojul** [reply]
@davyike we downgraded to Node v20.15.1 and the issue is gone. It started happening when we switch to Node v22.11.0. We are not using Vercel either. We tried Node v20.16.0 sometime ago, but it contains a memory leak https://github.com/nodejs/node/issues/54274

**Comment by sannajammeh** [comment]
#73274 is in draft and fixes this issue.

**Comment by wyattjoh** [comment, marked as answer]
Resolved in https://github.com/vercel/next.js/pull/73274. Accessible in Next.js 14.2.20 and 15.0.4-canary.40+

**Reply by mitrvlr** [reply]
What Node.js version do you recommend?

**Reply by Milos-Zbiljic** [reply]
next.js 14.2.4 together with node 20.17 works without issues so far in my case. Everything higher then 20.17 breaks the app.

**Reply by rolandszoke-byborg** [reply]
Sadly, we are stuck at using 14.2.19 since 14.2.20 (and above) actually breaks our fetch deduplication if we have the same fetch in the page and a layout. The strange thing is that it only happens in production, I was not able to reproduce it locally. Our node version is 20.13.1 but also tried it with 20.18.1 and it has the same issue.

**Comment by Jalle19** [comment]
Any references to the actual bug in `undici` that's causing this? I can't seem to find anything.

**Comment by ThePr0grammer** [comment]
Error still persists even after upgrade for 14.2.20. We are using Vercel for deployment. 

`⨯ TypeError: Response.clone: Body has already been consumed.
    at webidl.errors.exception (node:internal/deps/undici/undici:3384:14)
    at _Response.clone (node:internal/deps/undici/undici:8883:31)
    at /var/task/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:181023
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  digest: '2800333212',
  page: '/en/investments'
}`

**Reply by wyattjoh** [reply]
We've confirmed with other users that upgrading to 14.2.20 has resolved their issues. Can you double check that the version upgrade was completed? If that's not the case can you share your source?

**Comment by adarsh-modanwal-amla** [comment]
TypeError: Response.clone: Body has already been consumed.
    at webidl.errors.exception (node:internal/deps/undici/undici:3384:14)
    at _Response.clone (node:internal/deps/undici/undici:8883:31)
    at C:\Users\adarsh.modanwal\Desktop\My-workspace\znode-webstore10x-page-builder\node_modules\next\dist\compiled\next-server\app-page.runtime.prod.js:12:181023
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    
     "next": "14.2.12",
     FROM node:20-alpine AS base
     local env : node: v20.18.0

    any leads?

**Reply by sannajammeh** [reply]
Upgrade to next 14.2.20.

**Reply by adarsh-modanwal-amla** [reply]
still the issue is persist.
any other leads?

**Reply by nass59** [reply]
I had the same issue in`"next": "14.2.13"` and `node: v20.18.0`
I upgraded in `"next": "14.2.23"`, it works fine now :)

**Comment by Mgonand** [comment]
Error still present in `14.2.23` (deployed in Vercel in production environment):
```
TypeError: Response.clone: Body has already been consumed.
    at webidl.errors.exception (node:internal/deps/undici/undici:3384:14)
    at _Response.clone (node:internal/deps/undici/undici:8883:31)
    at /var/task/node_modules/.pnpm/next@14.2.23_@opentelemetry+api@1.9.0_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:181241
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  digest: '4270320753'
}
```
We cannot reproduce it consistently, but the issue is definitely there. Some people say it’s easier to reproduce when multiple tabs are open. I encounter the error occasionally, but I haven’t found a way to reproduce it reliably.

Any suggestions to find out the source?

**Reply by donhessman** [reply]
@Mgonand what version of Node are you running? I believe this problem is worse in certain versions of Node 20.x 

I am hosting my app in Vercel and make heavy use of the fetch caching features in server actions (which is what I think kicks fetching over to Undici instead of native fetch).

Reverting to Node 18.x and upgrading to next 14.2.23 seemed to silence pretty much all of my issues like this. This week we upgraded to Node 22.x in Vercel and the logs are looking good so far.

**Reply by Mgonand** [reply]
We went all out and upgraded to Next 15. Looking good so far, lets see how it goes.

**Comment by kurkRedBrain** [comment]
I'm assuming this issue is not fixed if you're on next 14 with node 22.x? as I have encountered this in the vercel build logs and locally.

**Comment by Clementhec** [comment]
Still not fixed I got the issue on 14.2.5

**Reply by grzegorzpokorski** [reply]
> Still not fixed I got the issue on 14.2.5

https://github.com/vercel/next.js/discussions/69635#discussioncomment-11466212

**Reply by Clementhec** [reply]
Oops, it works now ;) thanks grzegor

**Comment by judewang** [comment]
Still not fixed. I got the issue on v14.2.25 with Node.js v22 runtime on Vercel. Downgrade to Node.js v18 fixed the issue. I use parallel routes and some of the routes have the same fetch function.

**Comment by hq-aw** [comment]
We still have the issue on v14.2.24. It started when we upgraded to Node.js v22.
It is hard to understand from the discussions if this is a Next.js issue, a Node.js issue or an undici issue.

I couldn't find any discussions on this topic neither on Node.js nor undici.

**Reply by justin-mclaren** [reply]
It's a Next.js issue specific to how they're extending the fetch API to provide caching features.

**Reply by hq-aw** [reply]
@wyattjoh, could you please re-open this issue? This seems to still be affecting a lot of people.

**Comment by Mehdi40** [comment]
Hello,

I'm also encountering this issue on v14.2.24 _(Node.js v20)_.
In my case, the issue seems to come from next-axiom _(I guess it's when next-axiom clone the request, when you set **logRequestDetails** to true)_.
It seems to come from the **withAxiomRouteHandler()** function, setting the **requestDetails** const which call **requestToJSON()** with the request. That's what I think from my investigations so far.

The bug appears randomly on several API routes on our app, it's really difficult to isolate.

**Comment by KasperrrH** [comment]
Any updates on this? I’m still encountering this issue on Next.js v14.2.30 with Node.js v20.19.4. I read it should work with Node v20.17, but I can’t downgrade for security reasons. I’m planning to upgrade to Next v15 and Node v22 to try and resolve the issue, but I’m not sure if that will work. I could use some confirmation or advice.

The bug appears on a certain API route on the website

**Comment by snyamathi** [comment]
I think I've identified the underlying cause of the `TypeError: Response.clone: Body has already been consumed` and have a minimal reproduction at https://github.com/snyamathi/nextjs-mem-leak/blob/main/index.mjs

However, since Next now uses a custom `cloneResponse`, this should not be happening anymore, at least by the same mechanism. I think there is a memory leak in the `cloneResponse` implementation though.

More info in this PR https://github.com/vercel/next.js/pull/82678 which has some background info on the changes to Undici, how that caused the issue here, and why it was reolved.
